src/Roslyn.Diagnostics.Analyzers/Core/SymbolIsBannedAnalyzer.cs

```diff
@@ -0,0 +1,231 @@
+        {
+            var additionalFiles = compilationContext.Options.AdditionalFiles;
+
+            if (!TryGetApiData(additionalFiles, compilationContext.CancellationToken, out ApiData apiData))
+            {
+                return;
+            }
+
+            if (!ValidateApiFiles(apiData, out List<Diagnostic> errors))
+            {
+                compilationContext.RegisterCompilationEndAction(context =>
+                {
+                    foreach (Diagnostic cur in errors)
+                    {
+                        context.ReportDiagnostic(cur);
+                    }
+                });
+
+                return;
+            }
+
+            var bannedApis = ImmutableHashSet.CreateBuilder<ISymbol>();
+            foreach (var apiString in apiData.ApiList)
+            {
+                var symbol = compilationContext.Compilation.GetTypeByMetadataName(apiString.Text);
+                if (!(symbol is null))
+                {
+                    bannedApis.Add(symbol);
+                }
+            }
+
+            if (bannedApis.Count > 0)
+            {
+                var symbolDisplayFormat = compilationContext.Compilation.Language == LanguageNames.CSharp
+                    ? SymbolDisplayFormat.CSharpShortErrorMessageFormat
+                    : SymbolDisplayFormat.VisualBasicShortErrorMessageFormat;
+                compilationContext.RegisterOperationAction(
+                    oac => AnalyzeOperation(oac, bannedApis.ToImmutable(), symbolDisplayFormat),
+                    OperationKind.ObjectCreation,
+                    OperationKind.Invocation,
+                    OperationKind.EventReference,
+                    OperationKind.FieldReference,
+                    OperationKind.MethodReference,
+                    OperationKind.PropertyReference);
+            }
+        }
+
+        private static void AnalyzeOperation(OperationAnalysisContext oac, ImmutableHashSet<ISymbol> bannedSymbols, SymbolDisplayFormat symbolDisplayFormat)
+        {
+            ITypeSymbol type = null;
+            switch (oac.Operation)
+            {
+                case IObjectCreationOperation objectCreation:
+                    type = objectCreation.Type.OriginalDefinition;
+                    break;
+
+                case IInvocationOperation invocation:
+                    type = invocation.TargetMethod.ContainingType.OriginalDefinition;
+                    break;
+
+                case IMemberReferenceOperation memberReference:
+                    type = memberReference.Member.ContainingType.OriginalDefinition;
+                    break;
+            }
+
+            if (!(type is null) && bannedSymbols.Contains(type))
+            {
+                oac.ReportDiagnostic(Diagnostic.Create(SymbolIsBannedRule, oac.Operation.Syntax.GetLocation(), type.ToDisplayString(symbolDisplayFormat)));
+            }
+        }
+
+        private static ApiData ReadApiData(string path, SourceText sourceText)
```

Comment:
Probably good to move all the public API file read/write methods into a separate utility type and share code, as I can see more analyzers in future tracking such semantic aspects for symbols (I can do it as a follow up change as well).